lib/src/back_end/solution_cache.dart

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) 2024, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import '../piece/piece.dart';
+import 'solution.dart';
+import 'solver.dart';
+
+/// Maintains a cache of [Piece] subtrees that have been previously solved.
+///
+/// If a given [Piece] has newlines before and after it, then (in most cases,
+/// assuming there are no other constraints), then the way it is formatted
+/// really only depends on its leading indentation and state. In that case, we
+/// can format that piece using a separate Solver and insert the results in any
+/// Solution that has that piece at that leading indentation.
+///
+/// This cache stores those previously formatted subtree pieces so that
+/// [CodeWriter] can reuse them across Solution[s].
+class SolutionCache {
+  final _cache = <_Key, Solution>{};
+
+  /// Looks up a previously cached solution for formatting [root] with leading
+  /// [indent].
+  ///
+  /// If found, returns the cached solution. Otherwise solves it, caches it,
+  /// and returns the result.
+  Solution find(int pageWidth, Piece root, State state, int indent) {
+    // See if we've already formatted this piece at this indentation.
+    var key = (root, state, indent: indent);
+    if (_cache[key] case var solution?) return solution;
+
+    var solver = Solver(this, pageWidth: pageWidth, leadingIndent: indent);
```

Comment:
[optional] These classes could be decoupled if the cache took a computation callback. Constructing an object with an argument of `this` always makes me nervous - and I think this one is relatively easy to avoid.

I also think it would be fine to inline this class to `Solver` - make the `_cache` field a `Map` and write the small `_cache[key] ??= Solver(_cache, ...).format`

This would also let us move the cache argument to a private constructor and let the root `Solver` create it's own cache to address Kallen's comment.

Reply:
See my reply to Kallen's comment. We specifically don't want the Solver to own its own cache because we want to share the cache across all Solvers.